Java第7回課題 REST API

aoha968 opened the pull request:
第7回課題 REST API
・PostmanからGET/POST/PATCH/DELETEのリクエストを受け付けるプログラムです。

プログラム仕様はREADME.md に記載しました。

aoha968 committed:
Java第7回課題 REST API

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,100 @@
+ * Rest API定義クラス
+ */
+@RestController
+@RequestMapping("/api/v1/employees")
```
パスにバージョンを含めるのgoodですね！
いい設計です。

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,100 @@
+     */
+    @GetMapping("{id}")
+    public Employee getEmployeeById(@PathVariable long id){
+        Optional<Employee> employee = employeeRepository.findById(id);
```
ロジックはなるだけControllerに持たせずなるだけServiceに寄せたいですね。

aoha968 replied:
[2236725](https://github.com/aoha968/Java2211/pull/8/commits/2236725ae7d2cd04e53f0795d43acb86e70f63df)
こちらでロジックをServiceに移動しました。

yoshi-koyama replied:
問題ないです！

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/exception/EmployeeExceptionHandler.java:
```diff
@@ -0,0 +1,96 @@
+
+// 例外処理を行うハンドラークラス
+@ControllerAdvice
+@RestController
```
このアノテーションは不要ですね、

aoha968 replied:
[2236725](https://github.com/aoha968/Java2211/pull/8/commits/2236725ae7d2cd04e53f0795d43acb86e70f63df)
こちらで不要なアノテーションを削除しました。

yoshi-koyama replied:
問題ないです。

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/model/Employee.java:
```diff
@@ -0,0 +1,32 @@
+@Entity
+@Table(name = "Employees")
+@Data
+public class Employee {
```
これは私の考えですがFormとエンティティは分けて定義した方が良いかと思います。

これはオニオンアーキテクチャなどアプリケーションのアーキテクチャの考え方に基づいています。
このPRではこのままで良いですが興味があれば相談ください。

説明をしようと思います。

aoha968 replied:
アーキテクチャについてあまり理解・意識せずに実装してしまいました。
おそらく、おっしゃっていることは、
Employee.java：form と entity と dto が混在している
これを、
formの場合：EmployeeForm
entityの場合：Employee
dtoの場合：EmployeeDto　　で分けた方がよいと言われているのかなぁと想像しました。

正直、漠然としてしか理解できてないので、詳しく知りたいと思います。

このPRは対応せずにこのままにします。

yoshi-koyama replied:
> formの場合：EmployeeForm
entityの場合：Employee
dtoの場合：EmployeeDto　　で分けた方がよいと言われているのかなぁと想像しました。

これであっています。
すくなくともFormはControllerのレイヤーでのみ登場するように実装したいですね。
依存関係で言うと
Controller/Form -> Service/Entity/Repositoryという依存関係はOKですが、
Controller/Form <- Service/Entity/Repositoryという依存関係は避けたいところです。

yoshi-koyama replied:
> このPRは対応せずにこのままにします。

承知しました！

aoha968 committed:
不要なアノテーション削除とemployeeRepository.findById(id)のロジックをServiveに移動

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    @GetMapping("{id}")
+    public Employee getEmployeeById(@PathVariable long id){
+        Optional<Employee> employeeId = service.getEmployeeId(id);
+        if(employeeId.isPresent()) {
+            // 指定したIDをもつユーザーデータがあればそのユーザーデータを返す
+            return employeeId.get();
+        }else{
+            // 指定したIDをもつユーザーデータがなければnullを返す
+            return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
+        }
```
例外をかえすのもServiceで実装したらよいかと思います。

yoshi-koyama replied:
よくよく見ると、この処理は
```suggestion
return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
だけでよいですね。

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    @PostMapping
+    public Employee createEmployee(@Valid @RequestBody Employee employee) {
+        if(employeeRepository.countByEmailId(employee.getEmailId()) == 0){
+            // メールアドレスが未登録であれば登録する
+            return employeeRepository.save(employee);
+        }else{
+            // 登録済みであればエラーを返す
+            return service.getEmployeeId(employee.getId()).orElseThrow(() -> new ResourceNotFoundException("Employee is registered email:" + employee.getEmailId()));
```
メールアドレスが未登録であれば登録、登録済みであればエラーも業務処理と考えてServiceに実装したいです。

yoshi-koyama replied:
改めて見ると気づいたのですが、
登録済みであればResourceNotFoundExceptionでなく重複を表す例外をthrowするのが適切かと思います。
別PRでもよいですが、必須で対応してほしいです 🙇

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    @PatchMapping("{id}")
+    public Employee updateEmployee(@PathVariable long id, @Valid @RequestBody Employee employee) {
+        Optional<Employee> employeeId = employeeRepository.findById(id);
+        if(employeeId.isPresent()) {
+            // 指定したIDをもつEmployeeデータがあれば更新する
+            employee.setId(id);
+            return employeeRepository.save(employee);
+        }else {
+            // 指定したIDをもつEmployeeデータがなければエラーを返す
+            return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
指定したIDに対して該当する従業員がいるかいないかにより処理が分岐するのは業務処理としてとらえてServiceに実装してもいいと思いました。

yoshi-koyama replied:
あくまでもControllerはFormの内容をServiceに渡して返された結果をレスポンスに必要な形に変換するところだけを担うようにしたいです。

yoshi-koyama replied:
Controllerでロジックをもたせてしまうとあとあとテストするときもちょっと面倒ですね。


ただ、各レイヤーにどういった責務をもたせるかははーさんが設計していいところです。
このPRでは一旦これで実装してあとからリファクタ対応としてレイヤーの責務を整理して実装を直してもいいと思います。

yoshi-koyama replied:
参考資料送っておきますね。
http://terasolunaorg.github.io/guideline/current/ja/Overview/ApplicationLayering.html

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/service/EmployeeService.java:
```diff
@@ -0,0 +1,18 @@
+
+@Service
+public class EmployeeService {
+    @Autowired
+    private EmployeeRepository employeeRepository;
```
コンストラクタインジェクションにするのが望ましいです！

yoshi-koyama replied:
これについても後ほど対応でも構いません。

aoha968 replied:
コンストラクタインジェクションにしました
[1b39999](https://github.com/aoha968/Java2211/pull/8/commits/1b39999c49c92007ac245daa0277303ec7e9aeb1)

yoshi-koyama replied:
問題ないです！

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+    private EmployeeService service;
+
+    /**
+     * Employeeデータリストを取得する.
+     * @return Employeeデータリスト
+     */
```
わかりやすくて良いコメント。

yoshi-koyama reviewed (commented):
改善点はありますが、対応必須でないものやあとからでもいいものなのでLGTM

yoshi-koyama commented on lecture07/springboot-backend/pom.xml:
```diff
@@ -0,0 +1,73 @@
+			<artifactId>annotations</artifactId>
+			<version>RELEASE</version>
+			<scope>compile</scope>
+		</dependency>
```
これは要らない依存関係かも？

aoha968 replied:
色々試しているうちに不要な依存関係が入ったようです。
削除しました。
[f260001](https://github.com/aoha968/Java2211/pull/8/commits/f2600011f3c927859d807f61de170f02e05299c2)

yoshi-koyama replied:
確認しました！

yoshi-koyama commented on lecture07/springboot-backend/pom.xml:
```diff
@@ -0,0 +1,73 @@
+		<java.version>17</java.version>
+	</properties>
+	<dependencies>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jpa</artifactId>
+		</dependency>
```
JPAはリッチで使いやすい分、扱いに注意しないとキャッシュなどの関係で意図しない結果が生まれてしまう危険性があります。
より軽量なSpring Data JDBCを使ってみるのもいいかもしれません。
参考: https://speakerdeck.com/rshindo/jsug-2019-01

個人的にはSpring Data JPA難しいなあという印象……
Entityのライフサイクルとか……
https://www.baeldung.com/jpa-entity-lifecycle-events

Javaに限らずなのですが、ORMは便利ですが、発行するSQLが明確な実装にするほうが私は安心します。

※はーさんのレベル的にはここまで突っ込んでもいいのですが、Spring Data JPAも普通に現場で使われているのでこのままでも大丈夫です。

TomooAkamatsu commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+    @PatchMapping("{id}")
+    public Employee updateEmployee(@PathVariable long id, @Valid @RequestBody Employee employee) {
+        Optional<Employee> employeeId = employeeRepository.findById(id);
+        if(employeeId.isPresent()) {
```
このif文の内容は1行で書けると思います！

TomooAkamatsu commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+    public void deleteEmployee(@PathVariable long id) {
+        Optional<Employee> employee = employeeRepository.findById(id);
+        // 指定したIDをもつEmployeeデータがあればそのユーザーデータを削除する
+        if(employee.isPresent()) {
```
ここもそうですね

Tomoo-Akamatsu commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    @GetMapping("{id}")
+    public Employee getEmployeeById(@PathVariable long id){
+        Optional<Employee> employeeId = service.getEmployeeId(id);
```
employeeId, getEmployeeId
中身はEmployeeそのものなので命名がややひっかかりました。

TomooAkamatsu replied:
employeeId, getEmployeeId
中身はEmployeeそのものなので命名がややひっかかりました。

Tomoo-Akamatsu commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+            return employeeRepository.save(employee);
+        }else {
+            // 指定したIDをもつEmployeeデータがなければエラーを返す
+            return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
ここも好みの問題だとは思いますが、
- nullと分かった上でorElseThrow
- 確実に例外を出して返すことはないのにreturn

というのが気になります

```
employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
employee.setId(id);
return employeeRepository.save(employee);
```
この三行でOKな気もするのですが、この書き方が適切かどうかは私では判断できないっす笑
ただ、私の現場では早期リターンは推奨されてますね。(この場合は早期例外??)

YudaiNoda3576 replied:
早期returnはよく見かけるものの、早期例外は見かけないっすね🤔
例外はあくまでも異常な処理の場合に行う処理なので、コードの構造上正常系の処理のあとに書いてあって欲しいという気持ちが個人的にはあります！
でも早期例外がアンチパターンかというと、上記の理由は早期returnにも当てはまるのでそうではない気が。。。
@yoshi-koyama さん的にはどう思われますか？？

Tomoo-Akamatsu commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+        if(employee.isPresent()) {
+            employeeRepository.deleteById(id);
+        }else{
+            employee.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
ここもひっくり返せばif文はいらないかなあ...と

TomooAkamatsu replied:
ここもひっくり返せばif文はいらないかなあ...と

YudaiNoda3576 commented on lecture07/springboot-backend/src/test/java/net/javaguides/springboot/SpringbokBackendApplicationTests.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
+class SpringbokBackendApplicationTests {
```
実装されていないテストなどPRのスコープ外のファイルはコミットしないようにすると、現場では喜ばれると思います！（次回以降意識していただけると良いかと思います:smile:）

YudaiNoda3576 reviewed (commented):
コメントしています〜

TomooAkamatsu commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+            return employeeRepository.save(employee);
+        }else {
+            // 指定したIDをもつEmployeeデータがなければエラーを返す
+            return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
(諸事情によりゆーだいさんのコメントと順番が入れ替わってます、すみません)

ここも好みの問題だとは思いますが、

- nullと分かった上でorElseThrow
- 確実に例外を出して返すことはないのにreturn
というのが気になります

```
employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
employee.setId(id);
return employeeRepository.save(employee);
```

この三行でOKな気もするのですが、この書き方が適切かどうかは私では判断できないっす笑
ただ、私の現場では早期リターンは推奨されてますね。(この場合は早期例外??)

yoshi-koyama replied:
> コードの構造上正常系の処理のあとに書いてあって欲しいという気持ちが個人的にはあります！

これは私も同意です。
コード全体で正常、異常の分岐順を揃えておくのはいい取り組みだと思います。

yoshi-koyama replied:
> 早期returnはよく見かけるものの、早期例外は見かけないっすね🤔

こちらですよね。
```java
return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
こちら確かにぱっとみ例外をthrowしているように見えますが、
`employeeId.orElseThrow`なので例外をthrowしているわけではないですね。
従業員が存在すればその従業員のデータをかえす、そうでなければ例外をthrowするというものなので今のままで良いと思います。

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    @GetMapping("{id}")
+    public Employee getEmployeeById(@PathVariable long id){
+        Optional<Employee> employeeId = service.getEmployeeId(id);
```
確かに気になりますね。
変数名はemployee
メソッド名はgetEmployeeById
がよさそう。

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+     * @return 更新したEmployeeデータ
+     */
+    @PatchMapping("{id}")
+    public Employee updateEmployee(@PathVariable long id, @Valid @RequestBody Employee employee) {
+        Optional<Employee> employeeId = employeeRepository.findById(id);
+        if(employeeId.isPresent()) {
+            // 指定したIDをもつEmployeeデータがあれば更新する
+            employee.setId(id);
+            return employeeRepository.save(employee);
+        }else {
+            // 指定したIDをもつEmployeeデータがなければエラーを返す
+            return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
こちらの処理ですが、
```java
Optional<Employee> employeeId = employeeRepository.findById(id);
if(employeeId.isPresent()) {
    // 指定したIDをもつEmployeeデータがあれば更新する
    employee.setId(id);
    return employeeRepository.save(employee);
}else {
     // 指定したIDをもつEmployeeデータがなければエラーを返す
    return employeeId.orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```
下記のように書いたほうがよいかと思います。
```java
Optional<Employee> employeeId = employeeRepository.findById(id);
if(employeeId.isPresent()) {
    // 指定したIDをもつEmployeeデータがあれば更新する
    employee.setId(id);
    return employeeRepository.save(employee);
}else {
    // 指定したIDをもつEmployeeデータがなければエラーを返す
    throw new ResourceNotFoundException("Employee not exist with id:" + id);
```
またはServiceに下記のように実装するのもありかと思います。
```java
public void save(long id, String firstName, String lastName, String email) {
        Employee employee = employeeRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Employee not exist with id:" + id));
        employee.update(firstName, lastName, email); // Employeeに更新メソッドを作成しておく
        employeeRepository.save(employee);
    }
```

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,103 @@
+@RequestMapping("/api/v1/employees")
+public class EmployeeController {
+
+    /** Employeeアクセス用リポジトリ */
+    @Autowired
+    private EmployeeRepository employeeRepository;
+    /** EmployeeService */
+    @Autowired
+    private EmployeeService service;
```
各レイヤーの依存関係は下記のようにしたいですね。
Controller -> Service -> Repository（interface）

どのレイヤーがどう依存し合うかのルールを決めることで、メンテナンスしやすいアプリケーションにしたいです。

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/exception/EmployeeExceptionHandler.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<EmployeeExceptionResponse> handleException(Exception ex) {
+        EmployeeExceptionResponse exceptionResponse = new EmployeeExceptionResponse(
```
レスポンスに例外のスタックトレースを出すのは避けたいです。
理由は、レスポンスを見ることでソースコードのどこに何が書かれているかわかってしまし、セキュリティ的な脆弱性になります。
HTMLで画面を作るアプリケーションでも例外の詳細を画面に出すのはNGなのと同じですね。

レスポンスには
- 何が原因なのか（クライアント側？サーバー側？）
- 直すとしたらどこを直すか？（バリデーションなら何の項目がNGか？サーバーエラーなら暫く待つとか）
そういった点がわかるとよいです。
<img width="500" alt="スクリーンショット 2022-12-19 11 28 03" src="https://user-images.githubusercontent.com/62045457/208335952-03c0c38d-7ff1-47c1-aa06-747f707f947a.png">

一方で例外ハンドラーでスタックトレースはログに出すというのをやるのはいいと思います。

aoha968 replied:
素直に削除でもよかったのですが、デバッグ用に使用できた方が便利なので処理を無効化した対応にしました。
[e6c84dc](https://github.com/aoha968/Java2211/pull/8/commits/e6c84dc72464ede0a694d8e7a9605108eb069c48)

yoshi-koyama replied:
承知しました。
デバッグ用ということであればログに出すほうがAPIの設計としては自然なので今度余裕があれば対応いただければと思います 🙇

aoha968 committed:
不要な依存関係を削除

aoha968 committed:
@ControllerAdviceを追加

aoha968 committed:
DEBUG ON時のみスタックトレースを出力するように変更

aoha968 committed:
コンストラクタインジェクションに変更

aoha968 committed:
Controller処理を見直し
[Controller]
 - メールアドレスの未登録/登録判定処理をServiceに移動
 - CRUD処理を見直しServiceに移動できる箇所は移動
 - コメントの修正
[Service]
 - getEmployeeId -> getEmployeeById に名称変更&例外処理を追加
 - registerEmployee関数を追加
[Exception]
 - 不要なimport処理を削除
 - 不要な型宣言を削除

aoha968 commented:
Controller、Serviceで色々指摘を受けたので、指摘を参考に全体を見直し＆修正しました。
[c592dcb](https://github.com/aoha968/Java2211/pull/8/commits/c592dcbf55166cdc56cd97d9af5269bc1f3e93fe)

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    @PatchMapping("{id}")
+    public Employee updateEmployee(@PathVariable long id, @Valid @RequestBody Employee updateEmployee) {
+        Employee employee = service.getEmployeeById(id);
+        if(employee != null) {
+            // 指定したIDをもつEmployeeデータがあれば更新する
+            updateEmployee.setId(id);
+            return employeeRepository.save(updateEmployee);
+        }else {
+            // 指定したIDをもつEmployeeデータがなければnullを返す
+            return null;
+        }
```
このロジックもServiceに書くのが望ましいです。

aoha968 replied:
Controller -> Service に移動しました。
[3e9e4f2](https://github.com/aoha968/Java2211/pull/8/commits/3e9e4f2663a808010c982ae630a603aca3f449f5)

yoshi-koyama replied:
👍

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,87 @@
+     */
+    @DeleteMapping("{id}")
+    public void deleteEmployee(@PathVariable long id) {
+        Employee employee = service.getEmployeeById(id);
+        // 指定したIDをもつEmployeeデータがあればそのユーザーデータを削除する
+        if(employee != null) {
+            employeeRepository.deleteById(id);
+        }
```
このロジックもServiceに書くのが望ましいです。

aoha968 replied:
Controller -> Service に移動しました。
[3e9e4f2](https://github.com/aoha968/Java2211/pull/8/commits/3e9e4f2663a808010c982ae630a603aca3f449f5)

yoshi-koyama replied:
👍

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,87 @@
+            updateEmployee.setId(id);
+            return employeeRepository.save(updateEmployee);
+        }else {
+            // 指定したIDをもつEmployeeデータがなければnullを返す
+            return null;
```
レスポンスボディがnullというのは設計として違和感があります。
更新、削除の場合は更新/削除したかどうかに関わらず204でレスポンスボディなしでもいいかもですね。

aoha968 replied:
そもそも見直したら、service.getEmployeeById(id) で存在しなかったら例外を投げるので、nullチェックが不要だと気づきました。
また、以下処理を追加しました。
＜更新処理＞
・すでに存在しているメールアドレスを更新しようとすると失敗する処理が抜けていたため追加しました。
＜削除処理＞
・成功した場合は204のレスポンスボディなしで返却（失敗はservice.getEmployeeById(id)で例外を投げています。）
[3e9e4f2](https://github.com/aoha968/Java2211/pull/8/commits/3e9e4f2663a808010c982ae630a603aca3f449f5)

yoshi-koyama replied:
👍

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,87 @@
+    @Autowired
+    private EmployeeRepository employeeRepository;
+    /** EmployeeService */
+    @Autowired
+    private EmployeeService service;
```
ここもコンストラクタインジェクションにしたいです。

aoha968 replied:
こちらで対応しました。
[3e9e4f2](https://github.com/aoha968/Java2211/pull/8/commits/3e9e4f2663a808010c982ae630a603aca3f449f5)

yoshi-koyama replied:
問題ないです！

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/controller/EmployeeController.java:
```diff
@@ -0,0 +1,87 @@
+@RequestMapping("/api/v1/employees")
+public class EmployeeController {
+
+    /** Employeeアクセス用リポジトリ */
+    @Autowired
+    private EmployeeRepository employeeRepository;
```
Controller -> Repositoryの依存関係は廃止したいです。

aoha968 replied:
Serviceに移動させたので、依存関係を廃止しました。
[3e9e4f2](https://github.com/aoha968/Java2211/pull/8/commits/3e9e4f2663a808010c982ae630a603aca3f449f5)

yoshi-koyama replied:
OKです！

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/service/EmployeeService.java:
```diff
@@ -0,0 +1,50 @@
+@Service
+public class EmployeeService {
+
+    /**
+     * コンストラクタインジェクション
+     * "@Autowired"は、コンストラクタが1件のみかつSpring4.3以上であれば省略可能だがつけておく
+     * フィールドにfinalを使用することができて不変性を担保できるため、推奨される
+     */
```
コンストラクタインジェクションを推奨する意味まで調べていただき感謝です！問題ないです！

aoha968 committed:
更新/削除処理をServiceに移動、Controller -> Repositoryの依存関係を廃止、コンストラクタインジェクションにできる箇所を対応

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/service/EmployeeService.java:
```diff
@@ -0,0 +1,92 @@
+     * @return Employeeデータ
+     */
+    public void deleteEmployee(long id) {
+        Employee employee = getEmployeeById(id);   // idに紐づくEmployeeがなければコール側でエラー返却してくれる
```
Serviceのメソッド間での依存関係は持たせないほうがよいかと思います。
なぜならEmployeeをIDに応じて取得するのとIDで指定したEmployeeを削除するのは別の業務処理（ユースケースともいう）と捉えたいからです。

したがって下記のように実装するのが良いと私は考えます。

```suggestion
        Employee employee = employeeRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Employee not exist width id:" + id));
```

aoha968 replied:
同じような処理を複数記載するのはどうかと思い、上記のようにしたのですが、たしかに依存関係をもつことはよくないですね。
それなら、おっしゃられている通り別の業務処理として捉えた方がスマートだと思いました。
[20a7bc9](https://github.com/aoha968/Java2211/pull/8/commits/20a7bc9d214eb5aad7e73f5d958521fc3285bf6b)

yoshi-koyama replied:
👍

yoshi-koyama commented on lecture07/springboot-backend/src/main/java/net/javaguides/springboot/service/EmployeeService.java:
```diff
@@ -0,0 +1,92 @@
+    public void deleteEmployee(long id) {
+        Employee employee = getEmployeeById(id);   // idに紐づくEmployeeがなければコール側でエラー返却してくれる
+        employeeRepository.deleteById(id);         // 指定したIDをもつEmployeeデータがあればそのユーザーデータを削除する
+        throw new ResponseStatusException(HttpStatus.NO_CONTENT);   // とりあえず204のレスポンスボディなしで返却
```
204 NO_CONTENTは正常に処理が行われたときなので例外として扱うのは避けたいです。
このメソッドの戻り値の型をResponseEntityに変更するなどして対応すればいいと思います。

```java
@DeleteMapping("/names")
public ResponseEntity<Void> delete() {
    // 削除処理の呼び出し
    return ResponseEntity.noContent().build();
}
```

aoha968 replied:
ご指摘通り、正常処理なのに例外を投げることに違和感を感じたため、例外を投げずに戻り値で204を返却するようにしました。
[20a7bc9](https://github.com/aoha968/Java2211/pull/8/commits/20a7bc9d214eb5aad7e73f5d958521fc3285bf6b)

yoshi-koyama replied:
👍

aoha968 committed:
正常時のに204(NO_CONTENT)を返却と他Serviceのメソッド間での依存関係を持たせないように修正

yoshi-koyama reviewed (approved):
LGTMです！
対応ありがとうございました！
